src/NuGet.Clients/NuGet.PackageManagement.VisualStudio/Utility/FileWatchers/UserConfigFileWatcher.cs

```diff
@@ -0,0 +1,102 @@
+// Copyright (c) .NET Foundation. All rights reserved.
+// Licensed under the Apache License, Version 2.0. See License.txt in the project root for license information.
+
+#nullable enable
+
+using System;
+using System.IO;
+using NuGet.Common;
+using NuGet.Configuration;
+
+namespace NuGet.PackageManagement.VisualStudio.Utility.FileWatchers
+{
+    /// <summary>Watches a user settings directory to detect changes to nuget.config and config/*.config.</summary>
+    internal class UserConfigFileWatcher : IFileWatcher
+    {
+        private bool _disposed;
+        private FileSystemWatcher _userSettingsWatcher;
+        private FileSystemWatcher _configDirectoryWatcher;
+        private FileStream _lockFile;
+
+        public UserConfigFileWatcher()
+            : this(NuGetEnvironment.GetFolderPath(NuGetFolderPath.UserSettingsDirectory))
+        {
+        }
+
+        public UserConfigFileWatcher(string userSettingsDirectory)
+        {
+            _disposed = false;
+
+            string configDirectory = Path.Combine(userSettingsDirectory, "config");
+            string lockFile = Path.Combine(configDirectory, ".lock");
+
+
+            // FileSystemWatcher doesn't handle directories being deleted then recreated. So, take advantage that
+            // Windows doesn't let directories to be deleted when a file in the directory is being used.
+            // Since the config directory is in the user settings directory, a single lock can prevent both from
+            // being deleted.
+            if (!Directory.Exists(configDirectory))
+            {
+                Directory.CreateDirectory(configDirectory);
+            }
+            _lockFile = new FileStream(lockFile, FileMode.OpenOrCreate, FileAccess.Read, FileShare.ReadWrite);
+
+            _userSettingsWatcher = new FileSystemWatcher(userSettingsDirectory, Settings.DefaultSettingsFileName);
+            _userSettingsWatcher.Created += OnFileSystemEvent;
+            _userSettingsWatcher.Changed += OnFileSystemEvent;
+            _userSettingsWatcher.Deleted += OnFileSystemEvent;
+            _userSettingsWatcher.Renamed += OnFileSystemEvent;
+            _userSettingsWatcher.EnableRaisingEvents = true;
+
+            _configDirectoryWatcher = new FileSystemWatcher(configDirectory);
+            _configDirectoryWatcher.Created += OnFileSystemEvent;
+            _configDirectoryWatcher.Changed += OnFileSystemEvent;
+            _configDirectoryWatcher.Deleted += OnFileSystemEvent;
```

Comment:
Should we monitor config directory given it couldn't be renamed or deleted because of lock file presence?

Reply:
This is monitoring files in the directory, rather than the directory itself (see the tests), but I notice now that I forgot to add a `*.config` filter on the watcher 🤦

Reply:
What kind of config files are we expecting here? Currently on my local `configDirectory = C:\Users\xxxx\AppData\Roaming\NuGet\config` is empty. Do we need to watch any config file here?
Do mean to watch userSettingsDirectory instead?

Reply:
same question & answer as https://github.com/NuGet/NuGet.Client/pull/4939#discussion_r1033228718.  Resolving this comment too.